java/code/src/com/suse/manager/utils/SaltUtils.java

```diff
@@ -1361,41 +1357,51 @@ private static void handleHardwareProfileUpdate(MinionServer server,
      * @param server server this package will be added to
      * @return the InstalledPackage object
      */
-    private static InstalledPackage createPackageFromSalt(String name, Pkg.Info info,
-            Server server) {
+    private static InstalledPackage createPackageFromSalt(String name, Pkg.Info info, Server server) {
 
-        String epoch = info.getEpoch().orElse(null);
-        String release = info.getRelease().orElse("0");
-        String version = info.getVersion().get();
-        PackageEvr evr = PackageEvrFactory
-                .lookupOrCreatePackageEvr(epoch, version, release);
+        String serverArchTypeLabel = server.getServerArch().getArchType().getLabel();
 
         InstalledPackage pkg = new InstalledPackage();
-        pkg.setEvr(evr);
-        pkg.setArch(PackageFactory.lookupPackageArchByLabel(info.getArchitecture().get()));
+        pkg.setEvr(parsePackageEvr(serverArchTypeLabel, info));
         pkg.setInstallTime(new Date(info.getInstallDateUnixTime().get() * 1000));
         pkg.setName(PackageFactory.lookupOrCreatePackageByName(name));
         pkg.setServer(server);
+
+        // Add -deb suffix to architectures for Debian systems
+        String pkgArch = info.getArchitecture().get();
+        if (serverArchTypeLabel.equals("deb")) {
+            pkgArch += "-deb";
```

Comment:
It is called per package returned from profile update, which would be in 1000s for each system. But I can't see how `StringBuffer` would improve this case. IIUC, it won't reflect any changes in the bytecode since this is not a loop by itself.

Reply:
Bytecode level and object utilization will surely be different, albeit slightly.

In the current version, we are creating an initial instance of `String` (the original `pkgArch` referencing it), then we throw it into the garbage collector when we replace it with a new one (that's what `+=` does, updating `pkgArch` to point to the new `String` instance). Bytes are copied from the original `String` to the new `String`.

`StringBuffer` is definitely the wrong class to use here, as it's thread safe so its methods are synchronized. Acquiring and releasing locks takes time, so it would be slower. There is a thread-unsafe variant, `StringBuilder`, so let's look at that.

We would create a throwaway `StringBuilder`, and that internally gets a character array buffer. Assuming the buffer is big enough, the initial `pkgArch` would be copied there, and then the `"-deb"` string. At the end of the day, yet another new `String` would be allocated when calling `StringBuilder.toString()`, all characters so far copied and `StringBuilder` goes to GC.

So:
 - current implementation: one new character array allocation (for the result `String`), two string copy operations (one for the original `pkgArch` and one for `"-deb"`
 - `StringBuilder`: two new character array allocations (one for the bugger, one for the result String - assuming buffer is created big enough), three string copy operations (original `pkgArch` to buffer, `"-deb"` to buffer, buffer to result `String`)

I would say, let's tweak this only if we get a real performance problem and only if we measure that the problem is here :wink: